internal/output/output_result.go

```diff
@@ -0,0 +1,538 @@
+// This structured format facilitates generating various output formats (e.g., table, HTML, etc.).
+func BuildOutputResults(vulnResult *models.VulnerabilityResults) OutputResult {
+	var ecosystemMap = make(map[string][]OutputSourceResult)
+	var resultCount OutputVulnCount
+
+	for _, packageSource := range vulnResult.Results {
+		// Temporary workaround: it is a heuristic to ignore installed packages
+		// which are already covered by OS-specific vulnerabilities.
+		// This filtering should be handled by the container scanning process.
+		// TODO(gongh@): Revisit this after container scanning supports comprehensive functionality.
+		if strings.Contains(packageSource.Source.String(), "/usr/lib/") {
+			continue
+		}
+
+		// Process vulnerabilities for each source
+		sourceResult := processOutputSource(packageSource)
+		ecosystemMap[sourceResult.Ecosystem] = append(ecosystemMap[sourceResult.Ecosystem], sourceResult)
+		resultCount = updateVulnCount(resultCount, sourceResult.VulnCount)
+	}
+
+	// Build the final result
+	return buildOutputResult(ecosystemMap, resultCount)
+}
+
+// buildOutputResult builds the final OutputResult object from the ecosystem map and total vulnerability count.
+func buildOutputResult(ecosystemMap map[string][]OutputSourceResult, resultCount OutputVulnCount) OutputResult {
+	var ecosystemResults []OutputEcosystemResult
+	var osResults []OutputEcosystemResult
+	for ecosystem, sources := range ecosystemMap {
+		ecosystemResult := OutputEcosystemResult{
+			Name:    ecosystem,
+			Sources: sources,
+		}
+
+		if isOSEcosystem(ecosystem) {
+			ecosystemResult.IsOS = true
+			osResults = append(osResults, ecosystemResult)
+		} else {
+			ecosystemResults = append(ecosystemResults, ecosystemResult)
+		}
+	}
+
+	// Sort ecosystemResults to ensure consistent output
+	slices.SortFunc(ecosystemResults, func(a, b OutputEcosystemResult) int {
+		return cmp.Compare(a.Name, b.Name)
+	})
+
+	// Sort osResults to ensure consistent output
+	slices.SortFunc(osResults, func(a, b OutputEcosystemResult) int {
+		return cmp.Compare(a.Name, b.Name)
+	})
+
+	// Add project results before OS results
+	ecosystemResults = append(ecosystemResults, osResults...)
+
+	isContainerScanning := false
+	layers := getAllLayerInfo(ecosystemResults)
+	if len(layers) > 0 {
+		isContainerScanning = true
+	}
+	vulnTypeCount := getOutputVulnTypeCount(ecosystemResults)
+
+	return OutputResult{
+		Ecosystems:          ecosystemResults,
+		VulnTypeCount:       vulnTypeCount,
+		VulnCount:           resultCount,
+		IsContainerScanning: isContainerScanning,
+		AllLayers:           layers,
+	}
+}
+
+// processOutputSource processes a single source (lockfile or artifact) and returns an SourceResult.
+func processOutputSource(packageSource models.PackageSource) OutputSourceResult {
+	var sourceResult OutputSourceResult
+	packages := make([]OutputPackageResult, 0)
+	packageMap := make(map[string]bool)
```

Comment:
s/packageMap/packageSet

Also, since this is a set, we can use the `map[string]struct{}` type